### PR TITLE
Introduce new stream interface

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description

Please include a summary of the change and which issue is fixed or implemented. Please also include relevant motivation and context (e.g. links, docs, tickets etc.).

List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?
* suite of automated tests
* e2e tests on MacBook
* e2e tests on GPU machine
* e2e tests on Jetson Nano

YOUR_ANSWER

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [x] Docs updated? What were the changes:
- docstrings of new api
- user-gude documentation
- examples

### CODE REVIEW

**Review by sberan — CHANGES_REQUESTED** [review]
I think this will be a huge improvement! I would like it to work similarly to the current `Stream` implementation if possible so most of my suggestions are around that.


I still need to test on my nano so will keep you posted with those results.

File: inference/core/interfaces/stream/inference_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                    frame_timestamp=timestamp, frame_id=frame_id
+                )
+                predictions = self._model.predict(preprocessed_image)
+                self._watchdog.on_model_postprocessing_started(
+                    frame_timestamp=timestamp, frame_id=frame_id
+                )
+                predictions = self._model.postprocess(
+                    predictions,
+                    preprocessing_metadata,
+                    confidence=0.5,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py]: 👍  good catch

File: inference/core/interfaces/stream/watchdog.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+                post_processing_latency=post_processing_latency,
+                model_latency=model_latency,
+                e2e_latency=e2e_latency,
+            )
+        )
+
+
+def events_compatible(events: List[Optional[ModelActivityEvent]]) -> bool:
+    if not all_not_empty(sequence=events):
+        return False
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/watchdog.py:257]: 👍

File: inference/core/interfaces/camera/video_source.py [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+        previous_state = self._state
+        self._change_state(target_state=StreamState.RUNNING)
+        if previous_state is StreamState.PAUSED:
+            self._playback_allowed.set()
+        if previous_state is StreamState.MUTED:
+            self._frames_buffering_allowed = True
+
+    def _set_file_mode_buffering_strategies(self) -> None:
+        if self._buffer_filling_strategy is None:
+            self._buffer_filling_strategy = BufferFillingStrategy.WAIT
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/camera/video_source.py]: We can discuss on that - I believe that default behaviour against video is to process each frame.

File: inference/core/interfaces/stream/inference_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    def _dispatch_inference_results(self) -> None:
+        while True:
+            inference_results: Optional[
+                Tuple[datetime, int, np.ndarray, dict]
+            ] = self._predictions_queue.get()
+            if inference_results is None:
+                break
+            timestamp, frame_id, frame, predictions = inference_results
+            try:
+                self._on_prediction(frame, predictions)
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py]: Here - simply we do introduce breaking change

File: inference/core/interfaces/stream/inference_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+PREDICTIONS_QUEUE_SIZE = 256
+RESTART_ATTEMPT_DELAY = 1
+
+
+class InferencePipeline:
+    @classmethod
+    def init(
+        cls,
+        api_key: str,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py]: 👍

File: inference/core/env.py [hunk 6]
```diff
@@ -110,6 +110,9 @@
 
 # Flag to enforce FPS, default is False
 ENFORCE_FPS = str2bool(os.getenv("ENFORCE_FPS", False))
+MAX_FPS = os.getenv("MAX_FPS")
+if MAX_FPS is not None:
+    MAX_FPS = int(MAX_FPS)
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/env.py:122]: I am sorry, this PR is already large - I am happy to help as part of another one.

File: inference/core/interfaces/camera/video_source.py [hunk 7]
```diff
@@ -0,0 +1,1015 @@
...
+    UpdateSeverity,
+    VideoFrame,
+)
+from inference.core.interfaces.camera.exceptions import (
+    EndOfStreamError,
+    SourceConnectionError,
+    StreamOperationNotAllowedError,
+)
+
+DEFAULT_BUFFER_SIZE = int(os.getenv("VIDEO_SOURCE_BUFFER_SIZE", "64"))
```
Comment (paulguerrie) [inline inference/core/interfaces/camera/video_source.py:25]: Should these configs live in `inference/core/env.py`?

File: inference/core/interfaces/camera/video_source.py [hunk 8]
```diff
@@ -0,0 +1,1015 @@
...
+                f"Cannot connect to video source under reference: {self._stream_reference}"
+            )
+        self._source_properties = discover_source_properties(stream=self._video)
+        self._video_consumer.reset(source_properties=self._source_properties)
+        if self._source_properties.is_file:
+            self._set_file_mode_consumption_strategies()
+        else:
+            self._set_stream_mode_consumption_strategies()
+        self._playback_allowed.set()
+        self._stream_consumption_thread = Thread(target=self._consume_video)
```
Comment (paulguerrie) [inline inference/core/interfaces/camera/video_source.py:526]: Should 'daemon=True' be set on this thread so it stops in the case of the main thread getting stopped?

File: inference/core/interfaces/stream/inference_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+        * INFERENCE_PIPELINE_RESTART_ATTEMPT_DELAY - delay for restarts on stream connection drop
+
+        Returns: Instance of InferencePipeline
+
+        Throws:
+            * SourceConnectionError if source cannot be connected at start, however it attempts to reconnect
+                always if connection to stream is lost.
+        """
+        if api_key is None:
+            api_key = os.environ.get(API_KEY_ENV_NAMES[0], None) or os.environ.get(
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py:129]: we could - I am just wondering if we should. The reasoning:
if we put a high-level configuration option into the env - then the moment we latch the env variable into Py variable matters - if we latch the values in the moment we import the `env.py` - then any export to env that happens in between will not be visible and may lead to unexpected behaviour - like if someone has a script and config option is exported in the script function body after `import inference` - variable will not be visible for `inference` logic.

I am open to change that in a way you pointed, just bringing this into consideration.

File: inference/core/interfaces/stream/inference_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+            status_update_handlers=self._status_update_handlers,
+        )
+        logger.info(f"Inference thread started")
+        try:
+            for video_frame in self._generate_frames():
+                self._watchdog.on_model_preprocessing_started(
+                    frame_timestamp=video_frame.frame_timestamp,
+                    frame_id=video_frame.frame_id,
+                )
+                preprocessed_image, preprocessing_metadata = self._model.preprocess(
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py:238]: initially I was considering that, but the think convinced me not doing so was the metrics collected by watchodg:
```python
LatencyMonitorReport(
    frame_decoding_latency=0.024393796875000002, 
    pre_processing_latency=0.0021425156250000003, 
    inference_latency=0.033938140625000016, 
    post_processing_latency=0.0007386406250000003, 
    model_latency=0.03681929687500001, 
    e2e_latency=0.06121309375000001,
)
```
that represents avg processing time for 1080p@60fps stream against `yolov8n`.

Simply speaking preprocessing takes ~2ms and can be easily considered minor cost of the whole inference process. 
Post-processing - highly dependent on context (here I was inferring COCO-trained model against animal footage - low amount of bounding boxes was present) - can be indeed a bottleneck - let me check against another input

File: inference/core/interfaces/camera/video_source.py (hunk 7 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/camera/video_source.py:25]: 👍

File: inference/core/interfaces/camera/video_source.py (hunk 8 above)
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/camera/video_source.py:526]: hmm - good question
* on one hand - once our interface manages state of threads and join them (which is good) - we do not need
* on the other - users may left the threads hanging which would prevent the scripts termination - but if we are concerned about that - then we should probably apply to all cases, but then - threads that needs to dispose resources (like camera connection) may not be handled well

File: inference/core/interfaces/stream/inference_pipeline.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/interfaces/stream/inference_pipeline.py:238]: ok, I run the same test against video with noticeably more detections on average

```python
LatencyMonitorReport(
    frame_decoding_latency=0.049186390624999986, 
    pre_processing_latency=0.001831875, 
    inference_latency=0.033255812499999995, 
    post_processing_latency=0.00079621875, 
    model_latency=0.035883906250000014, 
    e2e_latency=0.08507029687499995,
)
```
this time, stream was 30fps, so the `frame_decoding_latency` is higher, but the post-processing is still sub 1ms

**Review by paulguerrie — APPROVED** [review]

**Pull request merged**